### ISSUE TITLE
[Feature] Metadata system for mods

### ISSUE THREAD

**Issue by Duckos-Mods** [UE4SS-RE/RE-UE4SS#258]
Currently there is no metadata system at all for mods. Some sort of metadata system would be useful.
This could allow users to specify stuff like the name of the main dll/script a mod loads something like `"main_file_name": "custom_mod_name"` Then instead of trying to load a file named `main.dll/lua` it would then try to load the file with the name the user specified.

This would also allow for a dependency system to utilize the metadata system 
```json
{
 "dependencies": [
  "ModNameOne",
  "ModNameTwo",
 ]
}
```
The dependency system should allow for version specifying or just the name and nothing more 

Specifying a version of the mod is also a valid thing `"version": [1, 0, 0]`

Allowing users to specify custom fields to read from could also be added. 
A full example of a metadata file could be something like 
```json
{
    "name": "MyAwsomeMod",
    "version": [1, 0, 0],
    "dependencies": {
        "OtherAwesomeMod": {
            "min": [1, 0, 0],
            "max": [2, 0, 0],
            "name": "OtherAwesomeMod"
        },
        "AnotherAwesomeMod": {
            "min": [1, 0, 0],
            "max": [0, 0, 0],
            "name": "AnotherAwesomeMod"
        }
    },
    "ue4ss_version": {
        "min": [2, 6, 0],
        "max": [0, 0, 0] // Allows any version above 2.6.0
    },
    "description": "This is my awesome mod!",
    "author": "UserName",
    "main_file_name": "MyCustomModMain"
}
```

**Comment by UE4SS** [comment]
This seems good but I think it's a good idea to put anything that only has one line like `description`,`author`, and `main_file_name` at the top of the file and anything that can have multiple lines at the bottom in any official examples.
That way you don't have to look through a sea of arrays and objects to find a single-line setting.

**Comment by narknon** [comment]
For C++ mods I'd like the metadata to be a struct in the base mod class.  That way the mod authors only need to edit it in one place, and glaze can generate a json for modmanagers to use.  Lua mod authors will unfortunately have to generate the metadata file themselves until we have a proper manager

**Comment by UE4SS** [comment]
> For C++ mods I'd like the metadata to be a struct in the base mod class. That way the mod authors only need to edit it in one place, and glaze can generate a json for modmanagers to use. Lua mod authors will unfortunately have to generate the metadata file themselves until we have a proper manager

I was thinking we would get rid of any metadata in the C++ mod class but I guess your way works as well.
We'd have to parse the metadata at runtime and re-set the metadata in the C++ mod class for each mod so that manual changes to the metadata json file actually have an effect.

**Comment by narknon** [comment]
Yeah, I was debating about that in my head, but I don't think end users should ever be editing the metadata manually (this also works as a bit of an anti-mod-theft mechanism).  My idea was to only reset the metadata if the mod's versioning was higher than the versioning in the metadata file.  So if mod is ver 0.2.0 and metadata is 0.1.0 it will replace it (potentially making a metadata.bak with the old one), but if both are 0.2.0 it will take the edits.

**Comment by narknon** [comment]
The more I think about it, the more I think there should only be one "true" metadata, and the file shouldn't really be edited by end users at all.  But, my idea above would still allow edits in an emergency where the mod author put a max version that isn't actually necessary or something.

**Comment by narknon** [comment]
My idea for handling it wouldn't have that issue.

**Comment by UE4SS** [comment]
> My idea for handling it wouldn't have that issue.

Then I don't understand your idea.

**Comment by narknon** [comment]
I also don't think it's common to include "max" versions at all.

**Comment by localcc** [comment]
Why bother with the min/max fields if we could just include semver there?

**Comment by narknon** [comment]
Version checks also don't have to be blocks. They can warn and prompt to enable anyway or disable.

**Comment by localcc** [comment]
imo if a version check fails it should block the mod from loading, as having an incorrect version will lead to unexpected behavior

**Comment by narknon** [comment]
I disagree, at least for ue4ss version.

**Comment by localcc** [comment]
why?

**Comment by narknon** [comment]
Because it's perfectly possible for a mod to work for several future versions of ue4ss

**Comment by localcc** [comment]
what about min version check? and it's not guaranteed for it to work with newer versions because stuff might and will change, which could break the mod

**Comment by localcc** [comment]
power users can edit the mod metadata in the case they want to see if it breaks

**Comment by narknon** [comment]
End users shouldn't have to go in and edit a bunch of mod files after updating ue4ss

**Comment by localcc** [comment]
end users shouldn't be launching mods that do not advertise compatibility for a ue4ss version, you are not talking about powerusers in this case, and if we allow loading incompatible mods, this will increase bugreport count from users loading mods on incompatible versions and potentially crashing, if we are talking about powerusers, they can edit the metadata

**Comment by Duckos-Mods** [comment]
I think at least if the version is to low it should throw an error where as if its to low it just throws an error. If its to high it throws a warning

**Comment by localcc** [comment]
but in both cases we could get a situation where an api was changed and will break the mod

**Comment by narknon** [comment]
Then they post the log and we see any warnings for out of date mods.

**Comment by localcc** [comment]
why should we by default allow users to crash their game and then waste both the user's and our time with figuring out the bug report?

**Comment by UE4SS** [comment]
Reference for semver: https://devhints.io/semver

**Comment by narknon** [comment]
You left that part out of your earlier statements.  But including a max that way would make it untested on all versions after when it was created and we wouldn't even get a warning that a mod was out of date so it's way worse for your argument that it would lead to unnecessary bug reports.

**Comment by localcc** [comment]
we should preferrably follow semver at least slightly, and not introduce breaking changes in at least patch releases, then it could be `~2.6.0`, which would allow loading on any `2.6.x` but not `2.7.x`. still, if anyone wants to load an incompatible mod, they should go and change the meta, adding functionality that allows doing this completely defeats the point of having ANY version checks

**Comment by localcc** [comment]
going and changing the meta is more of a pain than enabling a config option/etc., which makes it less likely that someone will actually do it to try and "fix" someone's mod

**Comment by narknon** [comment]
I'm saying there is no point in having ANY maximum UE4SS version checks.

**Comment by localcc** [comment]
then after we make a minor update with api changes mods just break and not notify the user they are incompatible, they just break and crash

**Comment by localcc** [comment]
how would they warn if there is no max version check?

**Comment by Duckos-Mods** [comment]
Is this now something worth working on considering it seems to have been discussed?

**Comment by nexustheru** [comment]
this can be done with pre defines. mods and setting has to put in certain values to get right sources and actor., when update the define only change inside ues, all user has to do is to add the variable. wich is the same.
